### PR TITLE
refactor(Observable): Subjects no longer wrapped in Subscriber

### LINKED ISSUE #748: Observable.subscribe checks 'instanceof' Subscription instead of shape
Currently `Subject` extends `Observable` and only implements `Subscription` so it doesn't see the `Subscription` function on it's inheritance chain.  This is an issue because currently the `subscribe` method on `Observable` checks whether the `subscriber` is an `instanceof Subscription`.  If the `subscriber` is a `Subject` it returns `false` and then wraps `subscriber` in a `Subscriber` and proceeds.  
```
      if (observerOrNext instanceof Subscriber) {
        subscriber = (<Subscriber<T>> observerOrNext);
      } else {
        subscriber = new Subscriber(<Observer<T>> observerOrNext);
      }
```
This seems faulty because this behavior will hide any attempt to override any of the `Subscriber` methods in the `Subject`. 
```
export class CustomSubject<T> extends Subject<T> {
  constructor(private _numObservables: number = 0) {
    super();
  }

  // Never gets called
  add(subscription?) {
    console.log("Add called")
  }
}
```
I was trying to create a custom `Subject` that reference counts the number of Observables it is subscribed to so it can defer calling `complete` until the last observable completes but this became an issue.  I don't know if this is a bug, but it seems like non-ideal behavior.  Would it be possible to add `Subscriber` to the prototypal chain of `Subject`, or,  in `Observable.subscribe` change the `instanceof` call to instead check the shape of the passed in `subscriber`?

### LINKED ISSUE #825: Using Subjects as Observers
I think this is a bug. Currently RxJS 5 Subjects cannot be used as Observers in `source.subscribe(subject)`, even though Subject implements the Observer interface.

```js
// Rx 5
observable.subscribe(
  subject.next.bind(subject),
  subject.error.bind(subject),
  subject.complete.bind(subject)
)

//Rx 4
observable.subscribe(subject.asObserver())
```

This should be simple to fix, but should be accompanied by tests too.

### PR DESCRIPTION
**Pull request by ntilwalli**
Observable.subscribe updated to check for instanceof Subject, test added, import order changed in Rx and KitchenSink to avoid circular dependency issue. closes #825, closes #748

### CODE REVIEW

**Comment by ntilwalli** [comment]
This PR addresses the issue where Subjects are unnecessarily wrapped in a Subscriber during Observable.subscribe.  Conceptually there is no reason they should be since they are designed to be both Observable and Subscriber.  This fix does introduce a circular dependency between Subject and Observable that must be managed (a bit inelegantly) during initial import in Rx and KitchenSink.  Is introducing module load ordering dependencies okay?  (Also after checkin I noticed I forgot to remove a commented out line of code in Observable.ts)

**Comment by blesh** [comment]
LGTM.

**Comment by blesh** [comment]
@ntilwalli can you rephrase this commit as `refactor(Observable): Subjects no longer wrapped in Subscriber` with a description of the work done, also add `closes #825, closes #748` to the commit message at the end.

The reason for the change from `fix` to `refactor`, is technically the functionality wasn't "broken" before, it was just sub-optimal.

Other than that, LGTM! Thanks for the PR!

File: src/Observable.ts [hunk 1]
```diff
@@ -89,7 +89,8 @@ export class Observable<T> implements CoreOperators<T>  {
     let subscriber: Subscriber<T>;
 
     if (observerOrNext && typeof observerOrNext === 'object') {
-      if (observerOrNext instanceof Subscriber) {
+      //if (observerOrNext instanceof Subscriber) {
```
Comment (kwonoj) [inline src/Observable.ts]: would you able to remove comment if this is leftover?

**Comment by kwonoj** [comment]
Except one minor comment about commented code, change looks good to me too.

**Comment by ntilwalli** [comment]
Great.  Done.

**Comment by ntilwalli** [comment]
Great.  Done.

**Comment by kwonoj** [comment]
@ntilwalli , would you able to flat 2 commit 42258b27b059f60eb5fb64d5532ef76be7e9f854 & 849d5b25b71e3e0e2743981965a504a5e2c5bbc1 into single one? both has identical commit message, and change seems quite similar too.

**Comment by staltz** [comment]
From issue #825, @blesh said:

> Subjects are being unnecessarily wrapped in Subscribers

I originally said the same. But after looking at code, I noticed that Subjects are **not** Subscribers. A Subject is an Observable and an Observer. And a Subscriber is an Observer. Every Observer is wrapped by a Subscriber when we subscribe with an observer. That's why a Subject is wrapped by a Subscriber.

Look at this inheritance diagram:

photo_2015-12-03_10-33-00

Yet this PR introduces:

```js
if (observerOrNext instanceof Subscriber || observerOrNext instanceof Subject) {
  subscriber = (<Subscriber<T>> observerOrNext);
}
```

Does TypeScript allow us to (unsafe?) cast `observerOrNext` to Subscriber if it's a Subject?

**Comment by kwonoj** [comment]
> Does TypeScript allow us to (unsafe?) cast...

: Since typescript only does type inference on build time, above code snippet could get passed.

`observerOrNext` is union type with primary type of `Observer<T>`, so regardless of runtime type checking (`observerOrNext instanceof ...`) typescript allows given parameter to be casted as `Subscriber`. Course, when code is being executed if assigned value is not `Subscriber`, it'll cause runtime issue.

if you're trying to cast `Subject` explicitly to `Subscriber` it'll complains, but also interestingly this snippet passes type inference.

```
let t:Subject<T> = ...;
subscribe(t);
```

while `t` is neither `Observer` or `((value: T) => void`.

Anyway, you're inheritance diagram seems legit, `Subject` seems does not inherit `Subscriber`.

**Comment by kwonoj** [comment]
Removed code snippet that I was thinking incorrectly.

**Comment by kwonoj** [comment]
> code is being executed if assigned value is not Subscriber, it'll cause runtime issue. 

: and seems this is not occurring since execution will be reach `add(...)` which is available implementation on `Subject` & `Subscriber` both.

**Comment by staltz** [comment]
Is reimplementing Subject as extending both Observable and Subscriber too much work? Since Subject implements both Observer and Subscription anyway.

**Comment by kwonoj** [comment]
I'm curious if subject need to extend subscriber. Nature of subject requires to be Observable and Observer as minimum - is there any specific usecases subscriber would benefit subject?

thinking about inheritance & implementation hierarchy, extending subscriber will makes subject

`Subject<T> extends Observable<T>, Subscriber<T> implements Observer<T>, Subscription<T>` (pseudo, let's forget about typescript doesn't support multiple inheritance. course in practical code it'll be more complex cause it need workaround for multiple inheritance..)

makes subject quite complex object.

**Comment by staltz** [comment]
> makes subject quite complex object.

I know, so maybe it's ok to keep it wrapped in a Subscriber.

**Comment by ntilwalli** [comment]
@staltz said:
>Is reimplementing Subject as extending both Observable and Subscriber too much work? Since Subject >implements both Observer and Subscription anyway.
What does *reimplementing* mean? Since Typescript is purely duck-typed, from a compile-time/type-checking/semantic  perspective, there's no difference between the statements:

1. `class Subject extends Observable<T> implements Observer<T>, Subscription<T>`
2. `class Subject extends Observable<T> implements Subscriber<T>`

Which means the cast is not unsafe. There *is* a difference between statements 1 and 2, and this:

3. `class Subject extends Observable<T> implements Observer<T>`

If the goal from from a semantic perspective is statements 1/2 then from a clarity perspective, 2 might be better and this PR is valid.  If the goal is 3 then this PR is not valid and there should be a PR to modify Subject accordingly.  When I first proposed this issue, I assumed Subject was coded correctly and Observable was inconsistent. It sounds like you're saying it's the other way around.

**Comment by ntilwalli** [comment]
I edited the above comment fix a reasoning error and to be more precise as to my intended meaning.

**Comment by blesh** [comment]
It will be fine. Are we tricking TypeScript? Yes. But we're doing it because TypeScript types aren't a real thing, they're helpful for build-time and development-time sanity checks. When you see `<casting>obj` anywhere, it's basically saying "I don't want your sanity checks, because I know what I'm doing". In this case, it works out. Subject and Subscriber have all of the same API points. This is by design.

This PR should be a small performance improvement when using Subjects to subscribe, and I don't see anything about it that could be bad. ... Other than people who _really_ like typing freaking out a little because we're being dirty, rotten liars about Subject being a Subscriber on that one line.

**Comment by ntilwalli** [comment]
@kwonoj I've flattened the commits.

**Comment by kwonoj** [comment]
@ntilwalli Thanks! I'll check in this after waiting couple of more hrs to see if there's any second opinion.

**Comment by staltz** [comment]
> This PR should be a small performance improvement when using Subjects to subscribe, and I don't see anything about it that could be bad. ... Other than people who really like typing freaking out a little because we're being dirty, rotten liars about Subject being a Subscriber on that one line.

Fair enough...

**Comment by kwonoj** [comment]
Merged with 5cb0f2b, thanks @ntilwalli !

**Comment by ntilwalli** [comment]
@kwonoj I just realized, when flattening the commits, I forgot to update the commit message to say "refactor" as @blesh asked.  It currently says "fix".  Should I amend my message and push it again?

**Comment by kwonoj** [comment]
Whew, sorry for that. I should've keep track detail before merge.. I think current branch is protected from force update, would like to hear @blesh which would be best.
